Mgv6 mudflow: Avoid floating stacked decorations

paramat opened the pull request:
Recently we started to remove decorations if the dirt below was flowed away,
but this did not check for stacked decorations, causing them to have only
their lowest node removed.

Remove nodes above a removed decoration until air is found.

Also remove 'old_is_water' bool which on testing is never true.
///////////////
Mgv6 mudflow: Add function to reduce indentation 
//////////////
 Mgv6 mudflow: Remove stacked decorations half-buried in placed mud.
Improve 'moveMud()' code and use index arguments instead of pointers.
//////////////

Fixes #5952 
Tested.

Since removing decorations above flowed and placed mud is only needed outside the mapchunk i have added conditions to only check outside, improving performace and so compensating for the checks for tall decorations.
'old_is_water' was a coding mistake, that bool is never true, so removed that check and action.

Mudflow comparison with master, to check unchanged behaviour:

![68747470733a2f2f692e696d6775722e636f6d2f5759726a5053662e6a7067](https://user-images.githubusercontent.com/3686677/27110133-04859562-509f-11e7-99c8-fb9e0ba0f979.jpeg)

^ Master

![screenshot_20170614_011325](https://user-images.githubusercontent.com/3686677/27110150-1cfca82e-509f-11e7-8264-7ff86941b6ac.png)

^ This PR

paramat commented:
Fixed, seems to work now.

Wuzzy2 commented:
It's not quite there:
![](https://i.imgur.com/TspWLak.png)

Note: All plants in this screenshot are supposed to have height 2. As you see, the plants near the cross hair have their upper part intact, while the bottom part is buried in grass blocks. xD

paramat commented:
Ok this is the other half of the problem, this time caused by the flowed mud that is being placed it replaces the lower node of your plants, we might need another check and removal.
This was probably happening before, maybe you didn't notice it?

paramat commented:
Added code cleanup commit.
However there is an issue with air columns in water, will fix.

paramat commented:
Air columns fixed with 3rd commit.

nerzhul commented on src/mapgen_v6.cpp:
```diff
@@ -915,6 +901,34 @@ void MapgenV6::flowMud(s16 &mudflow_minpos, s16 &mudflow_maxpos)
+			n3->getContent() != c_water_source) {
+		*n3 = MapNode(CONTENT_AIR);
+		// Check above for a stacked decoration and remove nodes
+		u32 i4 = i3;
```
seem to be a useless variable as it just takes i3 as a value and one time

paramat replied:
Yes, i have improved the code now to not use i4.

nerzhul commented:
code style with the function is better, ty

paramat commented:
Updated 3rd commit to also remove half-buried decorations above placed mud, @Wuzzy2 could you retest?
Also gretal improved and cleaned up code of new function.

Wuzzy2 commented:
Nope:
![](https://i.imgur.com/Ua4EknV.png)

paramat commented:
Some of those may be above a cave exavation? What you need to look for is the top half of a decoration above grass.

Wuzzy2 commented:
Oops, my bad. The cave generator was still active. I have now disabled it and re-tested. I fail to find any evidence of broken or buried plants, their height is preserved. :+1:

![](https://i.imgur.com/MHZVT7G.png)

paramat commented:
Thanks for the thorough testing.
Added screenshots to first post to show PR has unchanged mudflow behaviour.

nerzhul commented:
code is good for me , if you are sure the feature is okay merge, if not, just keep the refactor and merge

paramat commented:
Thanks. will do some more testing to be sure, then merge.

nerzhul commented:
A must have if i can request is to rename variables in 
```
void MapgenV6::moveMud(u32 i, u32 i2, u32 i3, v2s16 p2d, v3s16 em)
```

with more understandable names to understand what is eash variable, it's very obscure for this function

paramat commented:
Hmm, that would require rewriting 'flowMud()', which then means cleaning up all of that function, a big job. I think i prefer to do that in my forthcoming cleanup of 'flowMud()', hope this is ok. But yes doing that at some point seems reasonable.

nerzhul commented:
no no i don't mean rewrite flowMud, just change variable names in moveMud function which are contextually not linked with anything separately, let flowMud as now

paramat commented:
I see, ok will do.

paramat commented:
Updated as requested.

paramat commented:
Retested.

paramat commented:
f55816f93a8d80371878a5292f4f432e90dcc582
